presto-main/src/main/java/io/prestosql/operator/AbstractTopNOperator.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.operator;
+
+import io.prestosql.memory.context.AggregatedMemoryContext;
+import io.prestosql.memory.context.LocalMemoryContext;
+import io.prestosql.spi.Page;
+import io.prestosql.spi.block.SortOrder;
+import io.prestosql.spi.type.Type;
+
+import java.util.Iterator;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Verify.verify;
+import static java.util.Collections.emptyIterator;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Returns the top N rows from the source sorted according to the specified ordering in the keyChannelIndex channel.
+ */
+public abstract class AbstractTopNOperator
+{
+    private final LocalMemoryContext localUserMemoryContext;
+
+    private GroupedTopNBuilder topNBuilder;
+    private Iterator<Page> outputIterator;
+
+    AbstractTopNOperator(
+            AggregatedMemoryContext aggregatedMemoryContext,
+            List<Type> types,
+            int n,
+            List<Integer> sortChannels,
+            List<SortOrder> sortOrders)
+    {
+        requireNonNull(aggregatedMemoryContext, "aggregatedMemoryContext is null");
+        this.localUserMemoryContext = aggregatedMemoryContext.newLocalMemoryContext(AbstractTopNOperator.class.getSimpleName());
+        checkArgument(n >= 0, "n must be positive");
+
+        if (n == 0) {
+            outputIterator = emptyIterator();
+        }
+        else {
+            topNBuilder = new GroupedTopNBuilder(
+                    types,
+                    new SimplePageWithPositionComparator(types, sortChannels, sortOrders),
+                    n,
+                    false,
+                    new NoChannelGroupByHash());
+        }
+    }
+
+    void addInput(Page page)
+    {
+        boolean done = topNBuilder.processPage(requireNonNull(page, "page is null")).process();
```

Comment:
`topNBuilder can be null here`  I believe this is expected to be handled by the caller, so maybe just add a `verify` call with a clean message